pkg/chaosctl/common/exec.go

```diff
@@ -0,0 +1,124 @@
+			outNs, errNs := nsEnterExec(ctx, err.Error(), pod, daemon, cmd, c)
+			if errNs == nil {
+				return outNs, nil
+			}
+			err = fmt.Errorf("%s\nnsenter also failed with: %s", err.Error(), errNs.Error())
+		}
+		return "", err
+	}
+
+	return out, nil
+}
+
+func exec(ctx context.Context, pod v1.Pod, daemon v1.Pod, cmd string, c *kubernetes.Clientset) (string, error) {
+	name := pod.GetObjectMeta().GetName()
+	namespace := pod.GetObjectMeta().GetNamespace()
+	// TODO: if `containerNames` is set and specific container is injected chaos,
+	// need to use THE name rather than the first one.
+	// till 20/11/10 only podchaos and kernelchaos support `containerNames`, so not set it for now
+	containerName := pod.Spec.Containers[0].Name
+
+	req := c.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(name).
+		Namespace(namespace).
+		SubResource("exec")
+
+	req.VersionedParams(&v1.PodExecOptions{
+		Container: containerName,
+		Command:   []string{"/bin/sh", "-c", cmd},
+		Stdin:     false,
+		Stdout:    true,
+		Stderr:    true,
+		TTY:       false,
+	}, kubectlscheme.ParameterCodec)
+
+	var stdout, stderr bytes.Buffer
+	exec, err := remotecommand.NewSPDYExecutor(config.GetConfigOrDie(), "POST", req.URL())
+	if err != nil {
+		return "", fmt.Errorf("error in creating NewSPDYExecutor: %s", err.Error())
+	}
+	err = exec.Stream(remotecommand.StreamOptions{
+		Stdin:  nil,
+		Stdout: &stdout,
+		Stderr: &stderr,
+	})
+	if err != nil {
+		return "", fmt.Errorf("error in streaming remotecommand: %s", err.Error())
+	}
+	if stderr.String() != "" {
+		return "", fmt.Errorf(stderr.String())
+	}
+	return stdout.String(), nil
+}
+
+func nsEnterExec(ctx context.Context, stderr string, pod v1.Pod, daemon v1.Pod, cmd string, c *kubernetes.Clientset) (string, error) {
+	cmdSubSlice := strings.Fields(cmd)
+	if len(cmdSubSlice) == 0 {
+		return "", fmt.Errorf("command should not be empty")
+	}
+	// To disable printing irrelevant log from grpc/clientconn.go
+	// See grpc/grpc-go#3918 for detail. Could be resolved in the future
+	grpclog.SetLoggerV2(grpclog.NewLoggerV2(ioutil.Discard, ioutil.Discard, ioutil.Discard))
+	pid, err := GetPidFromPod(ctx, pod, daemon)
+	if err != nil {
+		return "", err
+	}
+	switch cmdSubSlice[0] {
+	case "ps":
+		nsenterPath := "-p/proc/" + strconv.Itoa(pid) + "/ns/pid"
+		nsCmd := fmt.Sprintf("mount -t proc proc /proc && %s && umount proc", cmd)
+		newCmd := fmt.Sprintf("/usr/bin/nsenter %s -- /bin/bash -c '%s'", nsenterPath, nsCmd)
+		return exec(ctx, daemon, daemon, newCmd, c)
+	case "cat", "ls":
+		// we need to enter mount namespace to get file related infomation
+		// but enter mnt ns would prevent us to access `cat`/`ls` in daemon
+		// so use `nsexec` to achieve using nsenter and cat together
+		if len(cmdSubSlice) < 2 {
+			return "", fmt.Errorf("%s should have one argument at least", cmdSubSlice[0])
+		}
+		newCmd := fmt.Sprintf("/usr/local/bin/nsexec %s %s", strconv.Itoa(pid), cmd)
```

Comment:
Rather than judging through the command, it would be better to add an argument to the function.

Or add a `MntNs` function for `ProcessBuilder` in `pkg/bpm` and use it to build the command string.

Reply:
My bad logic here is not correct. I was trying to filter out commands `cat`/`ls` that didn't have arguments, which is meaningless. Still thank you for the idea, I'll add an argument parsing tool for `nsexec` later.